lib/src/main/kotlin/at/bitfire/synctools/icalendar/validation/ICalPreprocessor.kt

```diff
@@ -40,18 +50,50 @@ class ICalPreprocessor {
         FixInvalidDayOffsetPreprocessor()   // fix things like DURATION:PT2D
     )
 
+    /**
+     * Applies [streamPreprocessors] to a given [String] by calling `fixString()` on each of them.
+     */
+    @VisibleForTesting
+    fun applyPreprocessors(input: String): String {
+        var newString = input
+        for (preprocessor in streamPreprocessors)
+            newString = preprocessor.fixString(newString)
+        return newString
+    }
+
     /**
      * Applies [streamPreprocessors] to a given [Reader] that reads an iCalendar object
      * in order to repair some things that must be fixed before parsing.
      *
-     * @param original    original iCalendar object
-     * @return            the potentially repaired iCalendar object
+     * The original reader content is processed in chunks of [chunkSize] lines to avoid loading
+     * the whole content into memory at once. If the given [Reader] does not support `reset()`,
+     * the whole content will be loaded into memory anyway.
+     *
+     * @param original original iCalendar object
+     * @return The potentially repaired iCalendar object.
+     * If [original] supports `reset()`,  the returned [Reader] will be a [SequenceReader].
+     * Otherwise, it will be a [StringReader].
      */
-    fun preprocessStream(original: Reader): Reader {
-        var reader = original
-        for (preprocessor in streamPreprocessors)
-            reader = preprocessor.preprocess(reader)
-        return reader
+    fun preprocessStream(original: Reader, chunkSize: Int = 1_000): Reader {
+        val resetSupported = try {
+            original.reset()
+            true
+        } catch(_: IOException) {
+            // reset is not supported. String will be loaded into memory completely
+            logger.warning("Reader does not support reset()")
+            false
+        }
+
+        if (resetSupported) {
+            val chunkedFixedLines = BufferedReader(original)
+                .lineSequence()
+                .chunked(chunkSize)
+                .map { chunk -> applyPreprocessors(chunk.joinToString("\n")) }
+            return SequenceReader(chunkedFixedLines)
```

Comment:
I think we can use Guava's `CharSource` like that:

~~~kotlin
BufferedReader(original).use { reader ->
val chunkedFixedLines = reader.lineSequence()
    .chunked(chunkSize)
    .map { chunk ->
        val fixed = applyPreprocessors(chunk.joinToString("\n"))
        CharSource.wrap(fixed)  // String to CharSource
    }.asIterable()
    return CharSource.concat(chunkedFixedLines).openStream()
}
~~~

Then we could avoid creating a custom reader.

But please double-check @ArnyminerZ, also that not all strings are kept in memory at once (but only one String at a time).

Reply:
I'd say this works, because the test that failed with huge iCalendars is passing.